### Clustering - Capstone Project
#### Please Use https://nbviewer.jupyter.org/ to copy paste notebook URL to view maps

## Segmenting and Clustering Neighbourhoods of the City of Toronto

In [1]:
import pandas as pd
import numpy as np
import scipy as sc
import matplotlib.pyplot as plt
import matplotlib as mpl
%matplotlib inline
import seaborn as sns


import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

!pip install pandas==1.0.3

import pandas as pd



#!pip install geocoder
#import geocoder # import geocoder

# initialize your variable to None
#lat_lng_coords = None

# loop until you get the coordinates
#while(lat_lng_coords is None):
#  g = geocoder.google('{}, Toronto, Ontario'.format(new_data['Postal Code']))
#  lat_lng_coords = g.latlng

#latitude = lat_lng_coords[0]
#longitude = lat_lng_coords[1]

print('Libraries imported.')

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.8.3
  latest version: 4.8.4

Please update conda by running

    $ conda update -n base -c defaults conda



# All requested packages already installed.

Libraries imported.


In [2]:
print('Hello Capstone Project Course!')

Hello Capstone Project Course!


## PART 1 - Building the Neighbourhood DataFrame

#### WEB SCRAPING

In [3]:
!pip install lxml

In [4]:
raw_data = pd.read_html("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M", header=0)
raw_data = raw_data[0]
raw_data.head(10)

Postal Code           Borough                                Neighbourhood
0         M1A      Not assigned                                 Not assigned
1         M2A      Not assigned                                 Not assigned
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
7         M8A      Not assigned                                 Not assigned
8         M9A         Etobicoke      Islington Avenue, Humber Valley Village
9         M1B       Scarborough                               Malvern, Rouge

#### Filtering only valid Boroughs - Removing "Not Assigned" Boroughs

In [5]:
new_data = raw_data[raw_data['Borough']!='Not assigned']
new_data.shape

(103, 3)

In [6]:
new_data.reset_index(drop=True,inplace=True)

In [7]:
new_data.head(12)

Postal Code           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   

                                        Neighbourhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                           Regent Park, Harbourfront  
3                    Lawrence Manor, Lawrence Heights  
4         Queen's Park, Ontario Provincial Government  
5             Islington Avenue, Humber Valley Village  
6                                      Malvern, Rouge  
7                                           Don Mills  
8                     Parkview Hill, Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park, Princess Gardens, Martin Grov...

In [8]:
new_data[new_data['Neighbourhood']=='Not Assigned']

# No Neighbourhoods Exist with NOT ASSIGNED if they have a borough

Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []

In [9]:
print("The Data Frame has {} rows and {} columns".format(new_data.shape[0], new_data.shape[1]))

The Data Frame has 103 rows and 3 columns


## PART 2 - Including the Latitude and Longitude

#### Getting Latitude and Longitude of Neighbourhoods - Google Geocoder is not complying - There is ARC GIS Option but not used

In [10]:
latlong = pd.read_csv('Geospatial_Coordinates.csv') 
latlong.head(5)

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

#### Joining Neighbourhoods with their Latitudes and Longitudes

In [11]:
neigh_with_latlong = new_data.merge(right=latlong,on='Postal Code')
neigh_with_latlong.head(12)

Postal Code           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   

                                        Neighbourhood   Latitude  Longitude  
0                                           Parkwoods  43.753259 -79.329656  
1                                    Victoria Village  43.725882 -79.315572  
2                           Regent Park, Harbourfront  43.654260 -79.360636  
3                    Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
5             Islington Avenue, Humber Valley Village  43.667856 -79.532242  
6                                      Malvern, Rouge  43.806686 -79.194353  
7                                           Don Mills  43.745906 -79.352188  
8                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
9                            Garden District, Ryerson  43.657162 -79.378937  
10                                          Glencairn  43.709577 -79.445073  
11  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724

In [12]:
# Getting Latitude and Longitude using ARC GIS

#!pip install geocoder
#import geocoder # Import geocoder package
#postal_code = new_data['Postal Code'] # Postal code for each neighborhood in Toronto, Canada

# Initialize your variable to 'None'
#lat_lng_coords = None

# Create an empty list to append the Latitude values
#lat_toronto = []

# Create an empty list to append the Longitude values
#lon_toronto = []

# Loop until getting the geographical coordinates
#for postal in postal_code:
#    g = geocoder.arcgis('{}, Toronto, Ontario'.format(postal))
#    lat_lng_coords = g.latlng
#    lat_toronto.append(lat_lng_coords[0])
#    lon_toronto.append(lat_lng_coords[1])

#neigh_with_latlong = new_data.copy()

#neigh_with_latlong['Latitude'] = lat_toronto
#neigh_with_latlong['Longitude'] = lon_toronto
#neigh_with_latlong.head(12)

## Part 3 - Exploring and Segmenting Neighbourhoods and Venues

In [13]:
len(neigh_with_latlong['Borough'].unique())

10

#### > We have 10 unique boroughs

In [14]:
neigh_with_latlong['Borough'].value_counts()

North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
York                 5
East York            5
East Toronto         5
Mississauga          1
Name: Borough, dtype: int64

In [15]:
len(neigh_with_latlong['Neighbourhood'])

103

In [16]:
neigh_with_latlong[neigh_with_latlong[['Neighbourhood']].duplicated()]

Postal Code     Borough Neighbourhood   Latitude  Longitude
13         M3C  North York     Don Mills  43.725900 -79.340923
46         M3L  North York     Downsview  43.739015 -79.506944
53         M3M  North York     Downsview  43.728496 -79.495697
60         M3N  North York     Downsview  43.761631 -79.520999

#### > We have 99 unique neighbourhoods and 103 total neighbourhoods with different postal codes

In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>ny_explorer</em>, as shown below.

In [17]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="tor_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


#### lets visualize ALL Boroughs and Neighbourhoods around toronto - Using Follium - Please use NB VIEWER WEBSITE and copy paste the project link

In [18]:
# create map of New York using latitude and longitude values
map_tor = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neigh_with_latlong['Latitude'], neigh_with_latlong['Longitude'], neigh_with_latlong['Borough'], neigh_with_latlong['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_tor)  
    
map_tor

#### lets visualize ONLY Boroughs around Downtown toronto and nearby Boroughs - Using Follium - Please use NB VIEWER WEBSITE and copy paste the project link

In [19]:
toronto_df = neigh_with_latlong[neigh_with_latlong['Borough'].str.contains('Toronto')]
toronto_df.head(3)

Postal Code           Borough                                Neighbourhood  \
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
9         M5B  Downtown Toronto                     Garden District, Ryerson   

    Latitude  Longitude  
2  43.654260 -79.360636  
4  43.662301 -79.389494  
9  43.657162 -79.378937

In [20]:
# create map of New York using latitude and longitude values
map_tor = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_df['Latitude'], toronto_df['Longitude'], toronto_df['Borough'], toronto_df['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_tor)  
    
map_tor

### Exploring Venues around Downtown Toronto and Nearby Boroughs

In [21]:
#@hidden_cell
CLIENT_ID = '0IRSU1PXCTADSJ0GAWTDUVNWMXS352WY4JU4XJN4XBLW4BV1' # your Foursquare ID
CLIENT_SECRET = 'TCNAWWPCZ22X34PMOHGEA2O0S1BILYL3BS3TQRD1Y5YQDAUG' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT= 100
radius = 500

Defining a function that retrieves a list venues for each neighbourhood and creates a dataframe

In [22]:
def getNearbyVenues(postal, borough, names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for postal, borough, name, lat, lng in zip(postal, borough, names, latitudes, longitudes):
        
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            postal,
            borough,
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = [
                  'Postal Code',
                  'Borough',
                  'Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)


In [23]:
toronto_venues = getNearbyVenues(  postal = toronto_df['Postal Code'],
                                   borough = toronto_df['Borough'],
                                   names = toronto_df['Neighbourhood'],
                                   latitudes=toronto_df['Latitude'],
                                   longitudes=toronto_df['Longitude']
                                  )

#### The Compelete DataFrame containing Downtown Toronto Boroughs, Nearby Boroughs and their VENUES

In [24]:
print(toronto_venues.shape)
toronto_venues.head(5)

(1652, 9)


Postal Code           Borough              Neighbourhood  \
0         M5A  Downtown Toronto  Regent Park, Harbourfront   
1         M5A  Downtown Toronto  Regent Park, Harbourfront   
2         M5A  Downtown Toronto  Regent Park, Harbourfront   
3         M5A  Downtown Toronto  Regent Park, Harbourfront   
4         M5A  Downtown Toronto  Regent Park, Harbourfront   

   Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                43.65426               -79.360636        Roselle Desserts   
1                43.65426               -79.360636           Tandem Coffee   
2                43.65426               -79.360636  Cooper Koo Family YMCA   
3                43.65426               -79.360636          Impact Kitchen   
4                43.65426               -79.360636     Body Blitz Spa East   

   Venue Latitude  Venue Longitude       Venue Category  
0       43.653447       -79.362017               Bakery  
1       43.653559       -79.361809          Coffee Shop  
2       43.653249       -79.358008  Distribution Center  
3       43.656369       -79.356980           Restaurant  
4       43.654735       -79.359874                  Spa

In [25]:
len(toronto_venues[toronto_venues['Borough']=='Downtown Toronto']['Neighbourhood'].unique())

19

### Counting how many venues per each neighbourhood

In [26]:
toronto_venues.groupby('Neighbourhood').count()

Postal Code  Borough  \
Neighbourhood                                                              
Berczy Park                                                  56       56   
Brockton, Parkdale Village, Exhibition Place                 24       24   
Business reply mail Processing Centre, South Ce...           16       16   
CN Tower, King and Spadina, Railway Lands, Harb...           17       17   
Central Bay Street                                           69       69   
Christie                                                     18       18   
Church and Wellesley                                         77       77   
Commerce Court, Victoria Hotel                              100      100   
Davisville                                                   31       31   
Davisville North                                              8        8   
Dufferin, Dovercourt Village                                 15       15   
First Canadian Place, Underground city                      100      100   
Forest Hill North & West, Forest Hill Road Park               4        4   
Garden District, Ryerson                                    100      100   
Harbourfront East, Union Station, Toronto Islands           100      100   
High Park, The Junction South                                26       26   
India Bazaar, The Beaches West                               19       19   
Kensington Market, Chinatown, Grange Park                    74       74   
Lawrence Park                                                 4        4   
Little Portugal, Trinity                                     43       43   
Moore Park, Summerhill East                                   3        3   
North Toronto West, Lawrence Park                            23       23   
Parkdale, Roncesvalles                                       14       14   
Queen's Park, Ontario Provincial Government                  34       34   
Regent Park, Harbourfront                                    45       45   
Richmond, Adelaide, King                                    100      100   
Rosedale                                                      4        4   
Roselawn                                                      2        2   
Runnymede, Swansea                                           38       38   
St. James Town                                               87       87   
St. James Town, Cabbagetown                                  49       49   
Stn A PO Boxes                                               96       96   
Studio District                                              40       40   
Summerhill West, Rathnelly, South Hill, Forest ...           15       15   
The Annex, North Midtown, Yorkville                          20       20   
The Beaches                                                   5        5   
The Danforth West, Riverdale                                 42       42   
Toronto Dominion Centre, Design Exchange                    100      100   
University of Toronto, Harbord                               34       34   

                                                    Neighbourhood Latitude  \
Neighbourhood                                                                
Berczy Park                                                             56   
Brockton, Parkdale Village, Exhibition Place                            24   
Business reply mail Processing Centre, South Ce...                      16   
CN Tower, King and Spadina, Railway Lands, Harb...                      17   
Central Bay Street                                                      69   
Christie                                                                18   
Church and Wellesley                                                    77   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              31   
Davisville North                                                         8   

#### Lets Analyze Each Neighbourhood within Downtown Toronto Borough and other nearby Boroughs (Containing TORONTO)

In [28]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

print(toronto_onehot.shape)
toronto_onehot.head()

(1652, 236)


Neighbourhood  Afghan Restaurant  Airport  Airport Food Court  \
0  Regent Park, Harbourfront                  0        0                   0   
1  Regent Park, Harbourfront                  0        0                   0   
2  Regent Park, Harbourfront                  0        0                   0   
3  Regent Park, Harbourfront                  0        0                   0   
4  Regent Park, Harbourfront                  0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0                 0                   0              0          0           0   
1                 0                   0              0          0           0   
2                 0                   0              0          0           0   
3                 0                   0              0          0           0   
4                 0                   0              0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0           0       1     0    0                 0                   0      0   
1           0       0     0    0                 0                   0      0   
2           0       0     0    0                 0                   0      0   
3           0       0     0    0                 0                   0      0   
4           0       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Board Shop  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
0           0              0          0         0                     0   
1           0              0          0         0                     0   
2           0              0          0         0                     0   
3           0              0          0         0                     0   
4           0              0          0         0                     0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Building  Burger Joint  \
0               0        0                0         0             0   
1               0        0                0         0             0   
2               0        0                0         0             0   
3               0        0                0         0             0   
4               0        0                0         0             0   

   Burrito Place  Bus Line  Butcher  Café  Cajun / Creole Restaurant  \
0              0         0        0     0                          0   
1              0         0        0     0                          0   
2              0         0        0     0                          0   
3              0         0        0     0      

#### Lets Group by Neighbourhood - Taking the mean of the frequency of occurrence of each category - Reduce the DataFrame Size

In [29]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()

print(toronto_grouped.shape)
toronto_grouped.head()

(39, 236)


Neighbourhood  Afghan Restaurant  \
0                                        Berczy Park                0.0   
1       Brockton, Parkdale Village, Exhibition Place                0.0   
2  Business reply mail Processing Centre, South C...                0.0   
3  CN Tower, King and Spadina, Railway Lands, Har...                0.0   
4                                 Central Bay Street                0.0   

    Airport  Airport Food Court  Airport Lounge  Airport Service  \
0  0.000000            0.000000        0.000000         0.000000   
1  0.000000            0.000000        0.000000         0.000000   
2  0.000000            0.000000        0.000000         0.000000   
3  0.058824            0.058824        0.117647         0.176471   
4  0.000000            0.000000        0.000000         0.000000   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0          0.000000                  0.0           0.0       0.0     0.017857   
1          0.000000                  0.0           0.0       0.0     0.000000   
2          0.000000                  0.0           0.0       0.0     0.000000   
3          0.117647                  0.0           0.0       0.0     0.000000   
4          0.000000                  0.0           0.0       0.0     0.000000   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0    0.000000                  0.0               0.0                 0.0   
1    0.000000                  0.0               0.0                 0.0   
2    0.000000                  0.0               0.0                 0.0   
3    0.000000                  0.0               0.0                 0.0   
4    0.014493                  0.0               0.0                 0.0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop    Bakery  Bank       Bar  \
0         0.0000   0.017857         0.0    0.017857  0.035714   0.0  0.000000   
1         0.0000   0.000000         0.0    0.000000  0.041667   0.0  0.041667   
2         0.0625   0.000000         0.0    0.000000  0.000000   0.0  0.000000   
3         0.0000   0.000000         0.0    0.000000  0.000000   0.0  0.058824   
4         0.0000   0.000000         0.0    0.000000  0.000000   0.0  0.014493   

   Baseball Stadium  Basketball Stadium     Beach  Bed & Breakfast  Beer Bar  \
0               0.0            0.017857  0.017857              0.0  0.035714   
1               0.0            0.000000  0.000000              0.0  0.000000   
2               0.0            0.000000  0.000000              0.0  0.000000   
3               0.0            0.000000  0.000000              0.0  0.000000   
4               0.0            0.000000  0.000000              0.0  0.000000   

   Beer Store  Belgian Restaurant    Bistro  Board Shop  Boat or Ferry  \
0         0.0                 0.0  0.017857         0.0       0.000000   
1         0.0                 0.0  0.000000         0.0       0.000000   
2         0.0                 0.0  0.000000         0.0       0.000000   
3         0.0                 0.0  0.000000         0.0       0.058824   
4         0.0                 0.0  0.000000         0.0       0.000000   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0        0.0  0.000000                   0.0        0.017857   0.0000   
1        0.0  0.000000                   0.0        0.083333   0.0000   
2        0.0  0.000000                   0.0        0.000000   0.0625   
3        0.0  0.058824                   0.0        0.000000   0.0000   
4        0.0  0.000000                   0.0        0.000000   0.0000   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  Butcher  \
0         0.000000       0.0      0.000000       0.000000       0.0   0.0000   
1         0.000000       0.0      0.000000       0.041667       0.0   0.0000   
2         0.000000       0.0      0.000000       0.062500       0.0   0.0625   
3         0.000000       0.0      0.000000       0.000000      

#### Let's print each neighborhood along with the top 5 most common venues

In [30]:
num_top_venues = 5

for hood in toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
            venue  freq
0     Coffee Shop  0.07
1        Beer Bar  0.04
2            Café  0.04
3  Farmers Market  0.04
4     Cheese Shop  0.04


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.12
1  Breakfast Spot  0.08
2     Coffee Shop  0.08
3             Gym  0.04
4          Office  0.04


----Business reply mail Processing Centre, South Central Letter Processing Plant Toronto----
                  venue  freq
0           Yoga Studio  0.06
1               Brewery  0.06
2  Gym / Fitness Center  0.06
3         Garden Center  0.06
4                Garden  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0   Airport Service  0.18
1    Airport Lounge  0.12
2  Airport Terminal  0.12
3             Plane  0.06
4     Boat or Ferry  0.06


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.17
1   

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [31]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

# This function will take one row at a time, arranges the values in descending order, and then takes its index values, which are the categories names of those values but only the highest 10
# Returns the Top 10 categories and we put those names in different columns for each neighbourhood

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [32]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...           Yoga Studio   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café           Cheese Shop                Bakery   
1        Breakfast Spot           Coffee Shop                   Gym   
2                Garden            Skate Park    Light Rail Station   
3        Airport Lounge      Airport Terminal      Sculpture Garden   
4        Sandwich Place    Italian Restaurant                  Café   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0          Cocktail Bar              Beer Bar            Restaurant   
1             Pet Store               Stadium         Burrito Place   
2         Burrito Place        Farmers Market  Fast Food Restaurant   
3               Airport    Airport Food Court                   Bar   
4             Juice Bar       Bubble Tea Shop   Japanese Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0        Farmers Market    Seafood Restaurant                   Park  
1            Restaurant          Climbing Gym                 Office  
2               Butcher            Restaurant          Auto Workshop  
3       Harbor / Marina              Boutique    Rental Car Location  
4       Thai Restaurant      Department Store            Salad Place

## Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [33]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0,n_init=10).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:30] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 4, 0, 1, 0,
       0, 0, 0, 0, 1, 2, 0, 0], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [34]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.merge(right=neighborhoods_venues_sorted, on='Neighbourhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0           Coffee Shop   
1  43.662301 -79.389494               0           Coffee Shop   
2  43.657162 -79.378937               0        Clothing Store   
3  43.651494 -79.375418               0           Coffee Shop   
4  43.676357 -79.293031               0                   Pub   

  2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0                Bakery                  Café                   Park   
1           Yoga Studio                 Diner  Portuguese Restaurant   
2           Coffee Shop                  Café         Cosmetics Shop   
3                  Café          Cocktail Bar               Beer Bar   
4           Coffee Shop     Health Food Store                  Trail   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Breakfast Spot                   Pub               Theater   
1    Persian Restaurant                  Park    Mexican Restaurant   
2   Japanese Restaurant       Bubble Tea Shop      Ramen Restaurant   
3            Restaurant   American Restaurant             Gastropub   
4          Neighborhood           Yoga Studio   Distribution Center   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0           Yoga Studio          Dessert Shop             Shoe Store  
1    Italian Restaurant            Hobby Shop  General Entertainment  
2             Bookstore           Pizza Place                  Hotel  
3        Lingerie Store      Department Store               Creperie  
4    Dim Sum Restaurant                 Diner         Discount Store

Finally, let's visualize the resulting clusters

In [35]:
# create map
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="tor_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude


map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Examine Clusters

Now, We can examine each cluster and determine the discriminating venue categories that distinguish each cluster.

#### Cluster 1

In [36]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]].head(10)

Borough  Cluster Labels 1st Most Common Venue  \
0  Downtown Toronto               0           Coffee Shop   
1  Downtown Toronto               0           Coffee Shop   
2  Downtown Toronto               0        Clothing Store   
3  Downtown Toronto               0           Coffee Shop   
4      East Toronto               0                   Pub   
5  Downtown Toronto               0           Coffee Shop   
6  Downtown Toronto               0           Coffee Shop   
7  Downtown Toronto               0         Grocery Store   
8  Downtown Toronto               0           Coffee Shop   
9      West Toronto               0              Pharmacy   

  2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0                Bakery                  Café                   Park   
1           Yoga Studio                 Diner  Portuguese Restaurant   
2           Coffee Shop                  Café         Cosmetics Shop   
3                  Café          Cocktail Bar               Beer Bar   
4           Coffee Shop     Health Food Store                  Trail   
5                  Café           Cheese Shop                 Bakery   
6        Sandwich Place    Italian Restaurant                   Café   
7                  Café                  Park             Baby Store   
8                  Café                   Gym         Clothing Store   
9                Bakery               Brewery            Supermarket   

       5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0             Breakfast Spot                   Pub               Theater   
1         Persian Restaurant                  Park    Mexican Restaurant   
2        Japanese Restaurant       Bubble Tea Shop      Ramen Restaurant   
3                 Restaurant   American Restaurant             Gastropub   
4               Neighborhood           Yoga Studio   Distribution Center   
5               Cocktail Bar              Beer Bar            Restaurant   
6                  Juice Bar       Bubble Tea Shop   Japanese Restaurant   
7                       Bank           Candy Store    Italian Restaurant   
8                 Restaurant       Thai Restaurant                 Hotel   
9  Middle Eastern Restaurant                  Café           Music Venue   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0           Yoga Studio          Dessert Shop             Shoe Store  
1    Italian Restaurant            Hobby Shop  General Entertainment  
2             Bookstore           Pizza Place                  Hotel  
3        Lingerie Store      Department Store               Creperie  
4    Dim Sum Restaurant                 Diner         Discount Store  
5        Farmers Market    Seafood Restaurant                   Park  
6       Thai Restaurant      Department Store            Salad Place  
7            Restaurant    Athletics & Sports            Coffee Shop  
8                   Bar        Breakfast Spot           Concert Hall  
9                   Bar                  Bank              Pet Store

In [54]:
toronto_grouped_2 = toronto_merged.join(toronto_grouped.set_index('Neighbourhood'), on='Neighbourhood')
toronto_grouped_2.groupby('Cluster Labels').mean().iloc[:,2:].T.iloc[:,0].sort_values(ascending=False).head(10)

Coffee Shop           0.083266
Café                  0.054875
Restaurant            0.030474
Park                  0.027934
Italian Restaurant    0.023582
Pizza Place           0.022443
Pub                   0.021828
Sandwich Place        0.021501
Bar                   0.021371
Bakery                0.019256
Name: 0, dtype: float64

#### Cluster 2

In [38]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]].head(10)

Borough  Cluster Labels 1st Most Common Venue  \
29   Central Toronto               1                  Park   
33  Downtown Toronto               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
29                 Trail          Tennis Court          Dessert Shop   
33            Playground                 Trail      Department Store   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
29  Ethiopian Restaurant           Escape Room            Electronics Store   
33           Escape Room     Electronics Store  Eastern European Restaurant   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
29  Eastern European Restaurant   Dumpling Restaurant             Donut Shop  
33          Dumpling Restaurant            Donut Shop       Doner Restaurant

In [55]:
toronto_grouped_2 = toronto_merged.join(toronto_grouped.set_index('Neighbourhood'), on='Neighbourhood')
toronto_grouped_2.groupby('Cluster Labels').mean().iloc[:,2:].T.iloc[:,1].sort_values(ascending=False).head(10)

Park                   0.416667
Trail                  0.291667
Tennis Court           0.166667
Playground             0.125000
Distribution Center    0.000000
Dessert Shop           0.000000
Dim Sum Restaurant     0.000000
Diner                  0.000000
Discount Store         0.000000
Doner Restaurant       0.000000
Name: 1, dtype: float64

#### Cluster 3

In [40]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]].head(10)

Borough  Cluster Labels    1st Most Common Venue  \
19  Central Toronto               2  Health & Beauty Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19                Garden           Yoga Studio          Dessert Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
19  Ethiopian Restaurant           Escape Room     Electronics Store   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
19  Eastern European Restaurant   Dumpling Restaurant             Donut Shop

In [56]:
toronto_grouped_2 = toronto_merged.join(toronto_grouped.set_index('Neighbourhood'), on='Neighbourhood')
toronto_grouped_2.groupby('Cluster Labels').mean().iloc[:,2:].T.iloc[:,2].sort_values(ascending=False).head(10)

Health & Beauty Service        0.5
Garden                         0.5
Yoga Studio                    0.0
Dessert Shop                   0.0
Ethiopian Restaurant           0.0
Escape Room                    0.0
Electronics Store              0.0
Eastern European Restaurant    0.0
Dumpling Restaurant            0.0
Donut Shop                     0.0
Name: 2, dtype: float64

#### Cluster 4

In [42]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]].head(10)

Borough  Cluster Labels 1st Most Common Venue  \
21  Central Toronto               3                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
21                 Trail         Jewelry Store      Sushi Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
21          Dessert Shop           Escape Room     Electronics Store   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
21  Eastern European Restaurant   Dumpling Restaurant             Donut Shop

In [57]:
toronto_grouped_2 = toronto_merged.join(toronto_grouped.set_index('Neighbourhood'), on='Neighbourhood')
toronto_grouped_2.groupby('Cluster Labels').mean().iloc[:,2:].T.iloc[:,3].sort_values(ascending=False).head(10)

Park                           0.25
Trail                          0.25
Jewelry Store                  0.25
Sushi Restaurant               0.25
Dessert Shop                   0.00
Escape Room                    0.00
Electronics Store              0.00
Eastern European Restaurant    0.00
Dumpling Restaurant            0.00
Donut Shop                     0.00
Name: 3, dtype: float64

#### Cluster 5

In [44]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]].head(10)

Borough  Cluster Labels 1st Most Common Venue  \
18  Central Toronto               4    Dim Sum Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
18                  Park              Bus Line           Swim School   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
18           Yoga Studio           Escape Room     Electronics Store   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
18  Eastern European Restaurant   Dumpling Restaurant             Donut Shop

In [58]:
toronto_grouped_2 = toronto_merged.join(toronto_grouped.set_index('Neighbourhood'), on='Neighbourhood')
toronto_grouped_2.groupby('Cluster Labels').mean().iloc[:,2:].T.iloc[:,4].sort_values(ascending=False).head(10)

Dim Sum Restaurant             0.25
Park                           0.25
Bus Line                       0.25
Swim School                    0.25
Yoga Studio                    0.00
Escape Room                    0.00
Electronics Store              0.00
Eastern European Restaurant    0.00
Dumpling Restaurant            0.00
Donut Shop                     0.00
Name: 4, dtype: float64

## Conclusion 
### Based on the clusters we can see the following:
1. Cluster 1: Downtown style cafe shops and resturants
2. Cluster 2: Only Gardens
3. Cluster 3: Out of downtown Areas Asian style restaurants and PUBs
4. Cluster 4: Trails and Summer Camps Areas
5. Cluster 5: Parks and Trails